Providing background for Baremetal IPI based enhancements (#102)
openshift/enhancements: enhancements/baremetal/baremetal.md

servers.

1. Will enable the user to deploy a control plane with 3 master nodes.
2. Will enable the user to grow the cluster by dynamically adding worker
nodes.
3. Will enable the user to scale down the cluster by removing worker nodes.

### Implementation Details/Notes/Constraints

Baremetal IPI is integrated with OpenShift through the metal3.io [8] project.
Metal3.io is a set of Kubernetes controllers that wrap the OpenStack Ironic
project to provide Kubernetes native APIs for managing deployment and
monitoring of physical hosts.

The installer support for Baremetal IPI deployments is described in more detail
in [7]. The installer runs on a special "provisioning host" that needs to be
connected to both a "provisioning network" and an "external network". The
provisioning network is a dedicated network used just for the purposes of
configuring baremetal servers to be part of the cluster. The traffic on the
provisioning network needs to be isolated from the traffic on the external
network (hence 2 seperate networks.). The external network is used to carry
cluster traffic which which includes cluster control plane traffic, application
and data traffic.

Control Plane Deployment

> 1. A minimin Baremetal IPI deployment consists of 4 hosts, one to be used

Reviewer: This doc needs to take into account or reference the future path for installer - removing bootstrap nodes, configuring a full master on the first instance, replacing the hardcoded load balancers with service load balancers for the apiserver, etc.  we don’t want to have a doc that embeds point in time context, or fails to move in the direction of the platform.

Author: Agreed. Will reference other enhancement docs from this one. The plan is to keep this doc alive with updates whenever new features are being worked on. Also, as new enhancement requests are written for new features affecting the baremetal platform, the idea is that these new enhancements will point back to this for context.

Reviewer 2: @sadasu I see clayton's request as outstanding.  The load balancers in particular.  How are they being created?  By which actor?  When?

first as a provisioning host and later potentially re-purposed as a worker.
The other 3 make up the control plane. These 4 hosts need to be connected
to both the provisioning and external networks.

2. Installation can be kicked off by downloading and running

> "openshift-baremetal-install". This image differs from the "openshift-install"

Reviewer: This is a deficiency - how are you planning to fix this?

Reviewer 3: The reason we're not part of the main openshift-install binary is our dependency on libvirt libraries. One proposal we discussed previously is to use https://github.com/digitalocean/go-libvirt which just hits the RPC directly, but we'd have to spend the time and rewrite the libvirt terraform provider.

This isn't on our radar to fix in the near term. There's a spike on the installer team to investigate removing the bootstrap host which would make this easier. 

Do you want this documented here?

Author: @smarterclayton, as mentioned by @stbenjam removing the bootstrap node would remove our dependency on libvirt. That way baremetal IPI installs can be part of the Openshift installer. This is part of our road map. I have updated the document to reflect that.

I think this "enhancement" is essentially provides a background that reviewers of baremetal IPI related code patches need to get context for what they are reviewing. We will make sure to keep this document alive. For these reasons, I think it would be helpful to merge this document.

Reviewer 2: Where and how is this separate installer maintained?

Reviewer 3: It is a separate installer binary, it is part of the same repo. 

It is distributed as part of the baremetal-installer image (like libvirt-installer is distributed as part of libvirt-installer). It's extractable using `oc` by extracting the `openshift-baremetal-install` command.

@sadasu You may want to include this information here.

authors:
  - "@sadasu"
reviewers:
  - "@smarterclayton"
  - "@abhinavdahiya"
  - "@enxebre"
  - "@deads2k"
approvers:
  - "@abhinavdahiya"
  - "@smarterclayton"
  - "@enxebre"
  - "@deads2k"
creation-date: 2019-11-06
last-updated: 2019-11-06
status: implemented
---

# Adding Baremetal IPI capabilities to OpenShift

This enhancement serves to provide context for a whole slew of features
and enhancements that will follow to make Baremetal IPI deployments via
OpenShift a reality.

At the time of this writing, code for some of these enhancements have already
merged, some are in progress and others are yet to me implemented. References
to all these features in different stages of development will be provided
below.

## Release Signoff Checklist

- [ ] Enhancement is `implementable`
- [ ] Design details are appropriately documented from clear requirements
- [ ] Test plan is defined
- [ ] Graduation criteria for dev preview, tech preview, GA
- [ ] User-facing documentation is created in [openshift/docs]

## Summary

Baremetal IPI deployments enable OpenShift to enroll baremetal servers to become
Nodes that can run K8s workloads.

> The Baremetal Operator [1] along with other provisioning services (Ironic and

Reviewer 4: I don't think this syntax works for links.  I've suggested an alternative that should work.
```suggestion
The [Baremetal Operator](1) along with other provisioning services (Ironic and
```

Reviewer 4: Note that there are several instances of this in the doc

dependencies) run in their own pod called "metal3". This pod is deployed by the
Machine API Operator when the Platform type is `BareMetal`. The OpenShift
Installer is responsble for providing all the necessary configs required for
a successful deployment.

## Motivation

The motivation for this enhancement request is to provide a background for all the
the subsequent enhancement requests for Baremetal IPI deployments.

### Goals

> The goal of this enhancement request is to provide context for all the changes

Reviewer 2: I think this doc should answer
1. comparison of metal3 to the AWS cloud provider.
2. components outside the cluster that have to be provisioned outside of a kube control plane, that will be used from inside the kube control plane.  (High level with links is fine).
3. new components inside the cluster that will speak to the outside metal3 components.  (High level with links is fine)
4. how internal apiserver LB, external apiserver LB, and ingresscontroller LBs will be managed an provisioned.  This needs to be pretty explicit
5. the network section from https://github.com/openshift/installer/blob/master/docs/user/metal/install_ipi.md#network-requirements with clear labeling about what, if anything, is managed on the cluster
6. top level item for the bootstrap host that the installer will use.

I may edit this comment to add more items as I have questions.

turn relies on the Ironic service [3] to manage and provision baremetal
servers.

1. Will enable the user to deploy a control plane with 3 master nodes.
2. Will enable the user to grow the cluster by dynamically adding worker
nodes.
3. Will enable the user to scale down the cluster by removing worker nodes.

### Implementation Details/Notes/Constraints

Baremetal IPI is integrated with OpenShift through the metal3.io [8] project.
Metal3.io is a set of Kubernetes controllers that wrap the OpenStack Ironic
project to provide Kubernetes native APIs for managing deployment and
monitoring of physical hosts.

The installer support for Baremetal IPI deployments is described in more detail
in [7]. The installer runs on a special "provisioning host" that needs to be
connected to both a "provisioning network" and an "external network". The
provisioning network is a dedicated network used just for the purposes of
configuring baremetal servers to be part of the cluster. The traffic on the
provisioning network needs to be isolated from the traffic on the external
network (hence 2 seperate networks.). The external network is used to carry

> cluster traffic which which includes cluster control plane traffic, application

Reviewer 2: lift the network section from the install doc. It's nice and crisp.  Make a section for it here.

Reviewer 4: for reference, https://github.com/openshift/installer/blob/master/docs/user/metal/install_ipi.md

and data traffic.

Control Plane Deployment

1. A minimin Baremetal IPI deployment consists of 4 hosts, one to be used
first as a provisioning host and later potentially re-purposed as a worker.
The other 3 make up the control plane. These 4 hosts need to be connected
to both the provisioning and external networks.

2. Installation can be kicked off by downloading and running
"openshift-baremetal-install". This image differs from the "openshift-install"
binary only because libvirt is needs to be always linked for the baremetal
install. Removing a bootstrap node would remove the dependency on libvirt
and then baremetal IPI installs can be part of the normal Openshift installer.
This is in the roadmap for this work and being investigated.

> 3. The installer starts a bootstrap VM on the provisioning host. With other

Reviewer 2: when you describe what a provisioning host is (that's missing at the moment), indicate what needs to be installed to make this VM startable.

platform types supported by OpenShift, a cloud already exists and the installer
runs the bootstrap VM on the control plane of this existing cloud. In the case
of the baremetal platform type, this cloud does not already exist, so the
installer starts the bootstrap VM using libvirt.

4. The bootstrap VM needs to be connected to the provisioning network and so the
the network interface on the provisioning host that is connected to the
provisioning network needs to be provided to the installer.

> 5. The bootstrap VM must be configured with a special well-known IP within the

Reviewer 2: well-known like "always the same" or well-known like the cluster-admin informs the installer?

provisioning network that needs to provided as input to the installer.

> 6. The installer user Ironic in the bootstrap VM to provision each host that

Reviewer 2: when you describe components, list ironic.  Is this running a static pod, is it a daemon on the bootstrap VM, something else?  Does this mean the bootstrap VM has to live past installation?

makes up the control plane. The installer uses terraform to invoke Ironic API
that configures each host to boot over the provisioning network using DHCP
and PXE.

> 7. The bootstrap VM runs a DHCP server and responds with network infomation and

Reviewer 2: dhcp server to provide IPs for which network. How long does this DHCP server need to exist?  is it user configurable (IP allocation is often dictated).

of in the bootstarp VM.

5. The provisioning IP used to bring up worker nodes remains the same as the control
plane case and the provisoning network also remains the same. The installer also
provides with a DHCP range within the same network that the workers are assigned IP
addresses from.

6. The ignition configs for the worker nodes are as passed as user data in the config
drive. Just as in the control plane hosts, Ironic power cycles the hosts that boot
using the RHCOS image now in their local disk. The host then joins the cluster as a
worker.

Currently, there is no way to pass the provisioning config known to the installer to
metal3 that is responsible for provisioning the workers.

### Risks and Mitigations

Will be specified in follow-up enhancement requests mentioned above.

## Design Details

### Test Plan

True e2e and integration testing can happen only after implementation for
enhancement [2] lands. Until then, e2e testing is being performed with the
help of some developer scripts.

Unit tests have been added to MAO and the Installer to test additions
made for the Baremetal IPI case.

### Graduation Criteria

> Metal3 integration is in tech preview in 4.2 and is targetted for GA in 4.4.

Reviewer 4: should update the release target to GA in 4.6

Reviewer 4: also fix the "targetted" typo

```suggestion
Metal3 integration is in tech preview in 4.2 and is targeted for GA in 4.6.
```

Metal3 integration is currently missing an important piece to information on
the baremetal servers and ther provisioning environment. Without this, true
end to end testing cannot be performed in order to graduate to GA.

### Upgrade / Downgrade Strategy

> Metal3 integration is in tech preview n 4.2 and missing key pieces that allows

Reviewer 4: ```suggestion
Metal3 integration is in tech preview in 4.2 and missing key pieces that allows
```
